test/e2e/ingress.go

```diff
@@ -0,0 +1,311 @@
+	name := fmt.Sprintf("%v%d", appPrefix, i)
+	l := map[string]string{}
+
+	Logf("Creating svc %v", name)
+	svc := svcByName(name, httpContainerPort)
+	svc.Spec.Type = api.ServiceTypeNodePort
+	_, err := c.Services(ns).Create(svc)
+	Expect(err).NotTo(HaveOccurred())
+
+	Logf("Creating rc %v", name)
+	rc := rcByNamePort(name, 1, testImage, httpContainerPort, l)
+	rc.Spec.Template.Spec.Containers[0].Args = []string{
+		"--num=1",
+		fmt.Sprintf("--start=%d", i),
+		fmt.Sprintf("--prefix=%v", pathPrefix),
+		fmt.Sprintf("--port=%d", httpContainerPort),
+	}
+	_, err = c.ReplicationControllers(ns).Create(rc)
+	Expect(err).NotTo(HaveOccurred())
+}
+
+// gcloudUnmarshal unmarshals json output of gcloud into given out interface.
+func gcloudUnmarshal(resource, regex string, out interface{}) {
+	output, err := exec.Command("gcloud", "compute", resource, "list",
+		fmt.Sprintf("--regex=%v", regex), "-q", "--format=json").CombinedOutput()
+	if err != nil {
+		Failf("%v", err)
+	}
+	if err := json.Unmarshal([]byte(output), out); err != nil {
+		Failf("%v", err)
+	}
+}
+
+func checkLeakedResources() error {
+	msg := ""
+	// TODO: These are the main resources but we should really check them all.
+	beList := []compute.BackendService{}
+	gcloudUnmarshal("backend-services", "k8s-be-[0-9]+", &beList)
+	if len(beList) != 0 {
+		for _, b := range beList {
+			msg += fmt.Sprintf("%v\n", b.Name)
+		}
+		return fmt.Errorf("Found backend services:\n%v", msg)
+	}
+	fwList := []compute.ForwardingRule{}
+	gcloudUnmarshal("forwarding-rules", "k8s-fw-.*", &fwList)
+	if len(fwList) != 0 {
+		for _, f := range fwList {
+			msg += fmt.Sprintf("%v\n", f.Name)
+		}
+		return fmt.Errorf("Found forwarding rules:\n%v", msg)
+	}
+	return nil
+}
+
+var _ = Describe("GLBC", func() {
+	// These variables are initialized after framework's beforeEach.
+	var ns string
+	var client *client.Client
+	var responseTimes, creationTimes []time.Duration
+
+	framework := Framework{BaseName: "glbc"}
+
+	BeforeEach(func() {
+		// This test requires a GCE/GKE only cluster-addon
+		SkipUnlessProviderIs("gce", "gke")
+		framework.beforeEach()
+		client = framework.Client
+		ns = framework.Namespace.Name
+		Expect(waitForRCPodsRunning(client, "kube-system", "glbc")).NotTo(HaveOccurred())
+		Expect(checkLeakedResources()).NotTo(HaveOccurred())
+		responseTimes = []time.Duration{}
+		creationTimes = []time.Duration{}
+	})
+
+	AfterEach(func() {
+		framework.afterEach()
+		wait.Poll(lbPollInterval, lbPollTimeout, func() (bool, error) {
+			if err := checkLeakedResources(); err != nil {
+				Logf("Still waiting for glbc to cleanup: %v", err)
+				return false, nil
+			}
+			return true, nil
+		})
+
+		// In most cases slow loadbalancer creation/startup translates directly to
+		// GCE api sluggishness. However this might be because of something the
+		// controller is doing, eg: maxing out QPS by repeated polling.
+		sort.Sort(timeSlice(creationTimes))
+		perc50 := creationTimes[len(creationTimes)/2]
```

Comment:
It feels kinda weird to insert what is essentially a performance test into AfterEach(), which is intended primarily for teardown.  But I understand that it's difficult to put it in a real "It()" type test without either polluting the existing functionality test, or wasting time setting up a second load balancer just to do the performance test.  Perhaps a reasonable middle-ground is to factor this performance test code into en explicit test function, and perhaps use Expect() rather than your if statement?